### ISSUE TITLE
Inference does weird things when a classmethod cls argument is called.

### ISSUE THREAD

**Issue by pitust** [PyCQA/pylint#11101]
### Bug description

Here is a minimal reproducer:

```python
class A:
    def b(self):
        return 0
    @classmethod
    def c(cls):
        return cls(), 0
class D:
    def e(self):
        self.f = A.c()[0]
        self.f.b() # x.py:10:8: E1101: Instance of 'D' has no 'b' member; maybe 'f'? (no-member)
```

Pylint deduces that `self.f` is an instance of `D` not `A`, here. It seems to have something to do with a scope escape type issue? Where `cls` starts to refer to `D` instead of `A` for some reason.

This code doesn't fail:

```python
class A:
    def b(self):
        return 0
    @classmethod
    def c(cls):
        return (_:=cls()), 0
class D:
    def e(self):
        self.f = A.c()[0]
        self.f.b()

```

Probably because the assignment will cause more aggressive inference.

### Configuration

```ini

```

### Command used

```shell
pylint x.py
```

### Pylint output

```python
************* Module x
x.py:10:8: E1101: Instance of 'D' has no 'b' member; maybe 'f'? (no-member)
```

### Expected behavior

No error.

### Pylint version

```shell
pylint 4.0.5
astroid 4.2.0
Python 3.14.5 (main, May 31 2026, 17:34:14) [GCC 16.1.1 20260430]
```

### Additional dependencies

```python

```

**Comment by Synvoya** [comment]
Reproduced on `main` (pylint `4.1.0-dev0` @ `2bfab5f`, astroid `4.2.0b3`, Python 3.12.8).

This is an **astroid inference** bug — pylint's `no-member` check (`pylint/checkers/typecheck.py:1128`, `inferred = list(node.expr.infer())`) just reports what astroid infers. Inferring the RHS in isolation is correct, but inferring the *stored instance attribute* is wrong:

```python
import astroid
from astroid import nodes
mod = astroid.parse('''
class A:
    def b(self): return 0
    @classmethod
    def c(cls): return cls(), 0
class D:
    def e(self):
        self.f = A.c()[0]
        self.f.b()
''')
assign = next(mod.nodes_of_class(nodes.Assign))
print(next(assign.value.infer()))          # <Instance of .A>   correct
attr = [n for n in mod.nodes_of_class(nodes.Attribute) if n.attrname == 'b'][0]
print(next(attr.expr.infer()))             # <Instance of .D>   cls re-bound to D
```

So `cls()` inside `A.c` is evaluated with `cls` bound to the class that *performs the attribute store*, not the defining class. Minimizing the trigger — it needs all three of: (1) a **second, unrelated class** doing the store (a single class does not reproduce), (2) the classmethod result reached via a **subscripted container** (`return (cls(), 0)` or `return [cls()]`; a bare `return cls()` does not fire), and (3) storage on an **instance attribute** (a plain local `f = A.c()` does not reproduce). The classes need not be related — this fires even when they are completely unrelated:

```python
class Container:
    @classmethod
    def make(cls): return [cls()]
    def hello(self): return 0

class Widget:
    def build(self):
        self.thing = Container.make()[0]
        self.thing.hello()   # E1101: Instance of 'Widget' has no 'hello' member (no-member)
```

This looks related to the `super()`/inheritance classmethod-factory reports (#981, #8908, #9203), but it's distinct: there's no inheritance and no `super()` here, and `cls` is bound to a class with no relationship to the one defining the classmethod. Looks like it needs an astroid fix — probably the `Needs astroid update` label.

**Comment by elafonta** [comment]
I do not know if this is linked, but I had an issue similar to this (the `__new__` returns types that have members that the polymorphic class doesn't).
https://github.com/drivendataorg/cloudpathlib/issues/579
Shoot me down if it's not linked, I'll delete my comment.  I didn't want to create a new issue for something already tracked.
